engine/robot.py

```diff
@@ -330,8 +374,97 @@ def set_phase(self, new_phase):
         with open(CSV_PATH + '/phases.csv', 'a') as fd:
             fd.write(str(self.phase) + '\n')
 
-    def execute_avoid_obstacle(self):
-        # TODO: SET BACK TO ORIGINAL MISSION (TRAVERSE OR RETURN)
+    def track_obstacle(self):
+        # make this async so main algorithm keeps running
+        # assuming front sensor is mounted in the center x position (width/2)
+        condition = True
+        curr_ultrasonic_value = 0
+        counter = 0  # added for testing
+        while condition:
+            if self.is_sim:
+                ultrasonic_value_file = open(ROOT_DIR + '/tests/functionality_tests/csv/ultrasonic_values.csv',
+                                             "r")  # added for testing
+                content = ultrasonic_value_file.readlines()
+                line = content[counter]
+                counter += 1
+                curr_ultrasonic_value = float((''.join(line.rstrip('\n')).strip('()').split(', '))[0])
+                condition = (counter <= (len(content) - 1))
+                ultrasonic_value_file.close()
+            else:
+                curr_ultrasonic_value = self.front_ultrasonic.distance()
+            if (self.phase == Phase.TRAVERSE) or (self.phase == Phase.RETURN) or (self.phase == Phase.DOCKING) or (
+                    self.phase == Phase.AVOID_OBSTACLE):
+                if curr_ultrasonic_value < self.detect_obstacle_range:
+                    # didn't check whether we can reach goal before contacting obstacle because obstacle detection does
+                    # not detect angle, so obstacle could be calculated to be falsely farther away than the goal. Not
+                    # optimal because in cases, robot will execute boundary following when it can reach goal
+                    self.dist_to_goal = math.sqrt(
+                        (self.state[0] - self.goal_location[0]) ** 2 + (self.state[1] - self.goal_location[1]) ** 2)
+                    self.avoid_obstacle = True
+                    with open(ROOT_DIR + '/tests/functionality_tests/csv/avoid_obstacle_result.csv', 'a') as fd:
+                        fd.write("Avoid" + '\n')
+                else:
+                    self.avoid_obstacle = False
+                    with open(ROOT_DIR + '/tests/functionality_tests/csv/avoid_obstacle_result.csv', 'a') as fd:
+                        fd.write("Not Avoid" + '\n')
+            if curr_ultrasonic_value < 0:
+                self.set_phase(Phase.FAULT)  # value should not go below 0; sensor is broken
+                with open(ROOT_DIR + '/tests/functionality_tests/csv/avoid_obstacle_result.csv', 'a') as fd:
+                    fd.write("Fault" + '\n')
+
+        # time.sleep(10)  # don't hog the cpu
+
+    def execute_avoid_obstacle(self, dist_to_goal, prev_phase, init_threshold, goal_threshold):
```

Comment:
all of these parameters are initialized in the robot. Should we do self.parameter instead? esp bc self.dist_to_goal seems to be updated in track_obstacle?

Reply:
we need dist_to_goal to stay static, so we'll be taking it from the parameter. The other params are removed